File: 0746-min-cost-climbing-stairs/0746-min-cost-climbing-stairs.py
class Solution:
    def f(self,cost,i,dp):
        if i>=len(cost):
            return 0
        
        if dp[i]!=-1:
            return dp[i]
        
        dp[i]=cost[i]+min(self.f(cost,i+1,dp),self.f(cost,i+2,dp))
        return dp[i]

    def minCostClimbingStairs(self, cost: List[int]) -> int:
        dp=[-1]*(len(cost)+1)
        return min(self.f(cost,0,dp),self.f(cost,1,dp))
        

File: 0746-min-cost-climbing-stairs/0746-min-cost-climbing-stairs.cpp
class Solution {
public:
    int minCostClimbingStairs(vector<int>& cost) {
        int m1=cost[0];
        int m2=cost[1];

        for(int i=2;i<cost.size();i++){
            int current = min(m1,m2)+cost[i];
            m1=m2;
            m2=current;
        }

        return min(m1,m2);
    }
};